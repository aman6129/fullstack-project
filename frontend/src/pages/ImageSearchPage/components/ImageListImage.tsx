import React, { useState } from 'react';
import axios from 'axios';
import { useAuth0 } from "@auth0/auth0-react";
import { IconButton, Card, CardMedia, CardActions } from '@material-ui/core';
import { StarBorder } from '@material-ui/icons';

import GiphyDataType from '../../../types/GiphyDataType';

interface ImageListImageProps {
  alt: string;
  image: GiphyDataType
}

const ImageListImage: React.FC<ImageListImageProps> = ({ image, alt }) => {
  const { user } = useAuth0();
  const [saving, setSaving] = useState<boolean>(false);
  
  const favoriteImage = async () => {
    setSaving(true);
    const { type, id, slug, url, title, images } = image;
    const payload = {
      slug,
      title,
      image_type: type,
      external_id: id,
      external_url: url,
      embed_url: images.fixed_height.url,
    };
  
    try {
      await axios.post('/api/user_favorites', {
        owner_email: user.email,
        ...payload,
      });
    } catch(e) {
      console.error(e);
    } finally {
      setSaving(false);
    }
  }

  const src = image.images.fixed_height.url;
  return (
    <Card>
      <CardMedia src={src} alt={alt} component='img' />
      <CardActions>
        <IconButton 
          aria-label="favorite"
          color="primary"
          disabled={saving}
          onClick={favoriteImage}
        >
          <StarBorder />
        </IconButton>
      </CardActions>
    </Card>
  );
}

export default ImageListImage;
