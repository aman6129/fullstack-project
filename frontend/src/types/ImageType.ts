interface ImageType {
  external_id: string;
  external_url: string;
  image_type: string;
  slug: string;
  embed_url: string;
  title: string;
}

export default ImageType;
