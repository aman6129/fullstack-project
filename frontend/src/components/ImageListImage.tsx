import React, { useState } from 'react';
import axios from 'axios';
import { useAuth0 } from "@auth0/auth0-react";
import { IconButton, Card, CardMedia, CardActions } from '@material-ui/core';
import { StarBorder } from '@material-ui/icons';

import ImageType from '../types/ImageType';

interface ImageListImageProps {
  alt: string;
  image: ImageType;
}

const ImageListImage: React.FC<ImageListImageProps> = ({ image, alt }) => {
  const { user } = useAuth0();
  const { embed_url: src } = image;
  const [saving, setSaving] = useState<boolean>(false);
  
  const favoriteImage = async () => {
    setSaving(true);
    try {
      await axios.post('/api/user_favorites', {
        owner_email: user.email,
        ...image,
      });
    } catch(e) {
      console.error(e);
    } finally {
      setSaving(false);
    }
  }

  return (
    <Card>
      <CardMedia src={src} alt={alt} component='img' />
      <CardActions>
        <IconButton 
          aria-label="favorite"
          color="primary"
          disabled={saving}
          onClick={favoriteImage}
        >
          <StarBorder />
        </IconButton>
      </CardActions>
    </Card>
  );
}

export default ImageListImage;
