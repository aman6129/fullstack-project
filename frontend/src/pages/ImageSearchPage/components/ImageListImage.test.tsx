import React from 'react';
import ImageListImage from './ImageListImage';
import { render } from '@testing-library/react';
import ImageType from '../../../types/ImageType';

const testImage: ImageType = {
  external_id: 'id',
  external_url: 'url',
  image_type: 'gif',
  slug: 'a-slug',
  embed_url: 'embed_url',
  title: 'test-image'
};

it('renders correctly', () => {
  const tree = render(<ImageListImage image={testImage} alt='test' />)

  expect(tree).toMatchSnapshot();
});
