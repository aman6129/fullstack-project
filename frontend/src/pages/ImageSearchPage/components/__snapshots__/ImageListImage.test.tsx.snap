// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <img
          alt="test"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
          src="embed_url"
        />
        <div
          class="MuiCardActions-root MuiCardActions-spacing"
        >
          <button
            aria-label="favorite"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <img
        alt="test"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        src="embed_url"
      />
      <div
        class="MuiCardActions-root MuiCardActions-spacing"
      >
        <button
          aria-label="favorite"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
